Refactoring bytecode_module to use flatcc instead of flatbuffers C++.

benvanik opened the pull request:
This improves our validation, removes manually-synchronized redundancy
with the schema for function descriptors, and optimizes a few things like
redundant type resolution.

Required a fix to cc_embed_data; the way it stores files as strings does not
guarantee alignment. The Flatbuffers C++ API papers over this and allows
for unaligned accesses, but flatcc doesn't (which is good); now all files
will be pointer-aligned.

Progress on #898.

benvanik committed:
Building bytecode_module_def with flatcc.

benvanik committed:
Linking in the flatcc bytecode_module_def files.
This ensures they are correctly generated and included.
Future changes will switch to using them over the C++ API.

benvanik committed:
Refactoring bytecode_module to use flatcc instead of flatbuffers C++.
This improves our validation, removes manually-synchronized redundancy
with the schema for function descriptors, and optimizes a few things like
redundant type resolution.

ScottTodd reviewed (commented):
Code LGTM, once we've figured out how to add the flatcc dep. I'd prefer to sequence this change after the dependency is added and fully merged though.

benvanik committed:
Fix flatcc target name and binary path (#2599)

This commit also defines iree_host_build_flatcc_cli when not
cross compiling so we can depend on it in iree_flatbuffer_c_library.

Also copy flatcc_cli from the proper place. They install it
in the source directory...

Co-authored-by: Lei Zhang <antiagainst@google.com>
